contracts/minter/src/contract.rs

```diff
@@ -0,0 +1,602 @@
+            code_id: msg.sg721_code_id,
+            msg: to_binary(&Sg721InstantiateMsg {
+                name: msg.sg721_instantiate_msg.name,
+                symbol: msg.sg721_instantiate_msg.symbol,
+                minter: env.contract.address.to_string(),
+                config: msg.sg721_instantiate_msg.config,
+            })?,
+            funds: info.funds,
+            admin: Some(info.sender.to_string()),
+            label: String::from("Fixed price minter"),
+        }
+        .into(),
+        id: INSTANTIATE_SG721_REPLY_ID,
+        gas_limit: None,
+        reply_on: ReplyOn::Success,
+    }];
+
+    Ok(Response::new()
+        .add_attribute("action", "instantiate")
+        .add_attribute("contract_name", CONTRACT_NAME)
+        .add_attribute("contract_version", CONTRACT_VERSION)
+        .add_attribute("sender", info.sender)
+        .add_submessages(sub_msgs))
+}
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn execute(
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    msg: ExecuteMsg,
+) -> Result<Response, ContractError> {
+    match msg {
+        ExecuteMsg::Mint {} => execute_mint_sender(deps, env, info),
+        ExecuteMsg::UpdateStartTime(expiration) => {
+            execute_update_start_time(deps, env, info, expiration)
+        }
+        ExecuteMsg::UpdatePerAddressLimit { per_address_limit } => {
+            execute_update_per_address_limit(deps, env, info, per_address_limit)
+        }
+        ExecuteMsg::MintTo { recipient } => execute_mint_to(deps, env, info, recipient),
+        ExecuteMsg::MintFor {
+            token_id,
+            recipient,
+        } => execute_mint_for(deps, env, info, token_id, recipient),
+        ExecuteMsg::SetWhitelist { whitelist } => {
+            execute_set_whitelist(deps, env, info, &whitelist)
+        }
+    }
+}
+
+pub fn execute_set_whitelist(
+    deps: DepsMut,
+    _env: Env,
+    info: MessageInfo,
+    whitelist: &str,
+) -> Result<Response, ContractError> {
+    let mut config = CONFIG.load(deps.storage)?;
+    if config.admin != info.sender {
+        return Err(ContractError::Unauthorized(
+            "Sender is not an admin".to_owned(),
+        ));
+    };
+    config.whitelist = Some(deps.api.addr_validate(whitelist)?);
+    CONFIG.save(deps.storage, &config)?;
+
+    Ok(Response::default()
+        .add_attribute("action", "set_whitelist")
+        .add_attribute("whitelist", whitelist.to_string()))
+}
+
+pub fn execute_mint_sender(
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+) -> Result<Response, ContractError> {
+    let config = CONFIG.load(deps.storage)?;
+    let sg721_address = SG721_ADDRESS.load(deps.storage)?;
+    let action = "mint_sender";
+    let mut pub_mint: bool = false;
+
+    // check if a whitelist exists and not ended
+    // sender has to be whitelisted to mint
+    if let Some(whitelist) = config.whitelist {
+        let wl_config: WhitelistConfigResponse = deps
```

Comment:
You will need to run these queries a lot. Same for external contracts. 
Best to move these under a helpers in a package and use it https://github.com/CosmWasm/cw-plus/blob/main/packages/cw4/src/helpers.rs

Reply:
This is a good idea. An optimization for the future for sure.